snapcraft/commands/lxc

```diff
@@ -2,8 +2,11 @@
 set -eu
 
 # Re-exec outside of apparmor confinement
-if [ -d /sys/kernel/security/apparmor ] && [ "$(while read -r l; do echo "$l"; done < /proc/self/attr/current)" != "unconfined" ]; then
-    exec /usr/bin/aa-exec -p unconfined -- "$0" "$@"
+if [ -d /sys/kernel/security/apparmor ]; then
+  label="$(cat /proc/self/attr/current)"
+  if [ "$label" != "unconfined" ] && [ -n "${label##*(unconfined)}" ]; then
+    exec aa-exec -p unconfined -- "$0" "$@"
+  fi
```

Comment:
The `while read -r` dance and the fully qualified path were used here to minimize the overhead of the `lxc` command that is already slow due to snap confinement. If we could keep that while supporting the new unconfined label it would be nice.

Reply:
Can you explain a bit more about why there was a need for the `while` loop here? Was it to read the label without having to spawn a new subprocess?

Reply:
https://github.com/canonical/lxd-pkg-snap/pull/189/commits/1465c1d764ef67e4efec5fdb41affaa217b5a1f9 gives slightly better performance:

```
[amurray:~] $ time (i=0; while [ $i -lt 1000 ]; do label=$(while read -r l; do echo $l; done < /proc/self/attr/current); i=$((i+1)); done)

real	0m1.435s
user	0m0.762s
sys	0m0.710s
[amurray:~] $ time (i=0; while [ $i -lt 1000 ]; do label=$(cat /proc/self/attr/current); i=$((i+1)); done)

real	0m2.786s
user	0m1.250s
```

But we could do this even faster if the commands used `bash` instead of `sh` for the interpreter, then you could read it as just `label=$(</proc/self/attr/current)`:

```
[amurray:~] $ time (i=0; while [ $i -lt 1000 ]; do label=$(</proc/self/attr/current); i=$((i+1)); done)

real	0m0.010s
user	0m0.000s
sys	0m0.010s
```

Reply:
Indeed, `dash`'s limitation makes it annoying. Historically we've stuck with a POSIX shell but we should re-assess that decision. That's something I'll look into and deal with in a different PR if we make the switch. Thanks!

Reply:
@tomponline we briefly discussed and agreed to switching to `bash` for the snap but I just ran those quick tests and think we should stick to `dash` for now.

For some reason, I cannot reproduce Alex's finding with `bash` efficiently reading from `procfs` in one go.

```
# reading from procfs
$ time /bin/sh -c 'for _ in $(seq 10000); do label=$(while read -r l; do echo $l; done < /proc/self/attr/current); done'

real	0m2.465s
user	0m1.662s
sys	0m0.969s

$ time /bin/bash -c 'for _ in $(seq 10000); do label=$(</proc/self/attr/current); done'

real	0m4.370s
user	0m3.186s
sys	0m1.505s
```

Either way, I think most of the "overhead" comes from the shell startup cost:

```
# spawning a shell
$ time (for _ in $(seq 10000); do /bin/sh -c 'true'; done)

real	0m11.102s
user	0m8.060s
sys	0m3.902s


$ time (for _ in $(seq 10000); do /bin/bash -c 'true'; done)

real	0m15.712s
user	0m10.653s
sys	0m5.951s
```

Combining both spawning the shell and reading from `procfs` only widens the gap:

```
# spawning a shell and reading from procfs
$ time (for _ in $(seq 10000); do /bin/sh -c 'label=$(while read -r l; do echo $l; done < /proc/self/attr/current)'; done)

real	0m13.735s
user	0m10.532s
sys	0m4.411s

$ time (for _ in $(seq 10000); do /bin/bash -c 'label=$(</proc/self/attr/current)'; done)

real	0m20.139s
user	0m14.123s
sys	0m7.376s
```